### PR TITLE
Fix ACLiC on macOS Big Sur

### PR DESCRIPTION
**Pull request by FonsRademakers**

### CODE REVIEW

File: gui/gui/src/TGWindow.cxx [hunk 1]
```diff
@@ -117,14 +117,7 @@ TGWindow::~TGWindow()
...
 {
-   if (!name && gDebug > 0) {
-      // set default frame names only when in debug mode
-      TString wname = ClassName();
-      wname += "::" + fName;
-      gVirtualX->SetWindowName(fId, (char *)wname.Data());
-   } else {
-      gVirtualX->SetWindowName(fId, (char *)name);
-   }
+   gVirtualX->SetWindowName(fId, (char *)name);
```
Comment (pcanal) [inline gui/gui/src/TGWindow.cxx:120]: I am confused.  If SetWindowName does not on windows, why keep it at (for windows)? I.e. why does "just" removing the alternate spelling when gDebug>0 helps?

File: core/base/src/TSystem.cxx [hunk 2]
```diff
@@ -1839,6 +1839,46 @@ static bool R__MatchFilename(const char *left, const char *right)
...
+   mapfile = "/System/Library/dyld/dyld_shared_cache_x86_64.map";
+#elif __arm64__
+   mapfile = "/System/Library/dyld/dyld_shared_cache_arm64e.map";
+#else
+   #error unsupported architecture
+#endif
+   std::ifstream cacheMap;
+   std::string line;
+   cacheMap.open(mapfile.c_str());
+   if (cacheMap.is_open()) {
```
Comment (Axel-Naumann) [inline core/base/src/TSystem.cxx:1863]:
```suggestion
   if (std::ifstream cacheMap{mapfile}) {
      std::string line;
```

File: core/base/src/TSystem.cxx [hunk 3]
```diff
@@ -1839,6 +1839,46 @@ static bool R__MatchFilename(const char *left, const char *right)
...
+      while (getline(cacheMap, line)) {
+         if (line.find(lib, 0) != std::string::npos) {
+           cacheMap.close();
+           lib.ReplaceAll("/usr/lib/lib","-l");
+           lib.ReplaceAll(".dylib","");
+           // skip these Big Sur libs as we cannot link with them
+           if (lib == "-loah" || lib == "-lRosetta")
+              lib = "";
+           return true;
+         }
```
Comment (Axel-Naumann) [inline core/base/src/TSystem.cxx:1873]:
```suggestion
         if (line.find(lib) != std::string::npos) {
            lib.ReplaceAll("/usr/lib/lib","-l");
            lib.ReplaceAll(".dylib","");
            // skip these Big Sur libs as we cannot link with them
            if (lib == "-loah" || lib == "-lRosetta")
               lib = "";
            return true;
         }
```

File: core/base/src/TSystem.cxx [hunk 4]
```diff
@@ -1839,6 +1839,46 @@ static bool R__MatchFilename(const char *left, const char *right)
 #endif
 }
 
+
+////////////////////////////////////////////////////////////////////////////////
+/// Check if lib is in the dynamic linker cache, returns true if it is.
```
Comment (Axel-Naumann) [inline core/base/src/TSystem.cxx:1844]:
```suggestion
/// Check if lib is in the dynamic linker cache, returns true if it is, and if so,
/// modifies the library file name parameter `lib` from `/usr/lib/libFOO.dylib`
/// to `-lFOO` such that it can be passed to the linker.
```

File: core/base/src/TSystem.cxx [hunk 5]
```diff
@@ -3694,7 +3734,7 @@ int TSystem::CompileMacro(const char *filename, Option_t *opt,
    Bool_t collectingSingleLibraryNameTokens = kFALSE;
    for (auto tokenObj : *linkLibrariesNoQuotes.Tokenize(" ")) {
       singleLibrary = ((TObjString*)tokenObj)->GetString();
-      if (!AccessPathName(singleLibrary) || singleLibrary[0]=='-') {
+      if (!AccessPathName(singleLibrary) || R__LibExistsInDylibCache(singleLibrary) || singleLibrary[0]=='-') {
```
Comment (Axel-Naumann) [inline core/base/src/TSystem.cxx:3737]: Let's do the fast things first: 
```suggestion
      if (singleLibrary[0]=='-' || !AccessPathName(singleLibrary) || R__LibExistsInDylibCache(singleLibrary)) {
```

File: gui/gui/src/TGWindow.cxx (hunk 1 above)
Comment (Axel-Naumann) [reply gui/gui/src/TGWindow.cxx:120]: Let me rephrase this, @pcanal let me know if my wording still captures what you tried to express :-)

How can this fix anything, how was this failing? Was `ClassName()` crashing?

File: core/base/src/TSystem.cxx [hunk 6]
```diff
@@ -1839,6 +1839,46 @@ static bool R__MatchFilename(const char *left, const char *right)
...
+static bool R__LibExistsInDylibCache(TString &lib)
+{
+#if !defined(R__MACOSX)
+   return false;
+#else
+   std::string mapfile;
+#if __x86_64__
+   mapfile = "/System/Library/dyld/dyld_shared_cache_x86_64.map";
+#elif __arm64__
+   mapfile = "/System/Library/dyld/dyld_shared_cache_arm64e.map";
```
Comment (Axel-Naumann) [inline core/base/src/TSystem.cxx:1856]:
```suggestion
   const char *mapfile = nullptr;
#if __x86_64__
   mapfile = "/System/Library/dyld/dyld_shared_cache_x86_64.map";
#elif __arm64__
   mapfile = "/System/Library/dyld/dyld_shared_cache_arm64e.map";
```

No need to construct `std::string`s just to pass a string literal around.

File: core/base/src/TSystem.cxx [hunk 7]
```diff
@@ -1839,6 +1839,46 @@ static bool R__MatchFilename(const char *left, const char *right)
...
 
+
+////////////////////////////////////////////////////////////////////////////////
+/// Check if lib is in the dynamic linker cache, returns true if it is.
+/// This is a unique feature of macOS 11.
+
+static bool R__LibExistsInDylibCache(TString &lib)
+{
+#if !defined(R__MACOSX)
+   return false;
```
Comment (Axel-Naumann) [inline core/base/src/TSystem.cxx:1850]:
```suggestion
   (void) lib; // suppress warning
   return false;
```

**Review by Axel-Naumann — CHANGES_REQUESTED** [review]
Nice solution, Fons, congrats!

File: gui/gui/src/TGWindow.cxx (hunk 1 above)
Comment (FonsRademakers) [reply gui/gui/src/TGWindow.cxx:120]: The problem is that SetWindowName does work for top level windows but not internal ones on macOS (have not tested this on other platforms). Found this running:
gDebug = 1
.x hsimple.C++

Which then crashed in cocoa code trying to set a window name on an internal wid that does not accept this operation. The gDebug flag setting the Window name is wrong as in the routine we don't know if the window is top level or not. So instead of trying to find that out it's better not have gDebug change behavior beyond printing debug info.

Comment (Axel-Naumann) [reply gui/gui/src/TGWindow.cxx:120]: Thanks, Fons! Now, your change just calls `    gVirtualX->SetWindowName(fId, (char *)name); ` unconditionally. How can that prevent the crash you describe?

Comment (Axel-Naumann) [reply gui/gui/src/TGWindow.cxx:120]: OK I got it - it's a side effect: with your diff applied, a `name == nullptr` is still passed to `SetWindowName()`, while before, a non-null `name` was constructed. The first action in `TGCocoa::SetWindowName()` is to return early if `name == nullptr`, avoiding the `assert` at `CocoaPrivate.mm:107`: "GetDrawable, non-existing drawable requested" and, for debug builds, an error at like 97 and 103: "This drawable was deleted already", "This drawable not found among allocated/deleted drawables".

I.e. `TGWindow::fId` is wrong. It's `1`, coming from `gVirtualX->GetDefaultRootWindow()`, as part of `TGClient` calling `TGFrame()`. At that point, `CocoaPrivate::fDrawables` is empty, and so the window id 1 cannot be found.

I can actually reproduce the exact same issue with macOS 10.15. I'm fine with fixing this the way you suggest, but I'll add a comment to explain that this relies on `name == nullptr`.

Comment (Axel-Naumann) [reply gui/gui/src/TGWindow.cxx:120]:
```suggestion
#ifdef R__MACOSX
   // MacOS fails to find the drawable of GetDefaultRootWindow(), only subsequent
   // windows ids can be found. See discussion here:
   // https://github.com/root-project/root/pull/6757#discussion_r518776154
   if (fId == gVirtualX->GetDefaultRootWindow())
      return;
#endif

   if (!name && gDebug > 0) {
      // set default frame names only when in debug mode
      TString wname = ClassName();
      wname += "::" + fName;
      gVirtualX->SetWindowName(fId, (char *)wname.Data());
   } else {
      gVirtualX->SetWindowName(fId, (char *)name);
   }
```

**Review by Axel-Naumann — APPROVED** [review]

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on mac10beta/cxx17

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on mac10beta/cxx17

**Comment by Axel-Naumann** [comment]
@phsft-bot build just on mac10beta/cxx17

**Comment by Axel-Naumann** [comment]
Yay! 🥳
